<!DOCTYPE html>
<html lang="en-US">
    <head>
        <meta charset="utf-8"> 
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="author" content="seqyuan" />
        <meta name="copyright" content="seqyuan" />

        <meta property="og:type" content="article" />
        <meta name="twitter:card" content="summary">

<meta name="keywords" content="pseudotime， trajectory inference, single-cell, " />

<meta property="og:title" content="单细胞轨迹分析dyno使用教程 "/>
<meta property="og:url" content="/step by step dyno.html" />
<meta property="og:description" content="dyno--1个R包实现几十种单细胞轨迹推断分析使用教程" />
<meta property="og:site_name" content="seqyuan" />
<meta property="og:article:author" content="seqyuan" />
<meta property="og:article:published_time" content="2020-01-13T00:00:00+08:00" />
<meta name="twitter:title" content="单细胞轨迹分析dyno使用教程 ">
<meta name="twitter:description" content="dyno--1个R包实现几十种单细胞轨迹推断分析使用教程">

        <title>单细胞轨迹分析dyno使用教程  · seqyuan
</title>
        <link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/css/bootstrap-combined.min.css" rel="stylesheet">
        <link href="//netdna.bootstrapcdn.com/font-awesome/4.0.1/css/font-awesome.css" rel="stylesheet">
        <link rel="stylesheet" type="text/css" href="/theme/css/pygments.css" media="screen">
        <link rel="stylesheet" type="text/css" href="/theme/tipuesearch/tipuesearch.css" media="screen">
        <link rel="stylesheet" type="text/css" href="/theme/css/elegant.css" media="screen">
        <link rel="stylesheet" type="text/css" href="/theme/css/custom.css" media="screen">
    </head>
    <body>
        <div id="content-sans-footer">
        <div class="navbar navbar-static-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
                        <span class="icon-bar"></span>
                        <span class="icon-bar"></span>
                        <span class="icon-bar"></span>
                    </a>
                    <a class="brand" href="/"><span class=site-name>seqyuan</span></a>
                    <div class="nav-collapse collapse">
                        <ul class="nav pull-right top-menu">
                            <li ><a href="">Home</a></li>
                            <li ><a href="/categories.html">Categories</a></li>
                            <li ><a href="/tags.html">Tags</a></li>
                            <li ><a href="/archives.html">Archives</a></li>
                            <li><form class="navbar-search" action="/search.html" onsubmit="return validateForm(this.elements['q'].value);"> <input type="text" class="search-query" placeholder="Search" name="q" id="tipue_search_input"></form></li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <div class="row-fluid">
                <div class="span1"></div>
                <div class="span10">
<article>
<div class="row-fluid">
    <header class="page-header span10 offset2">
    <h1><a href="/step by step dyno.html"> 单细胞轨迹分析dyno使用教程  </a></h1>
    </header>
</div>

<div class="row-fluid">
        <div class="span8 offset2 article-content">

            
            <p>在上一期的《<a href="https://mp.weixin.qq.com/s/6ZVTR3wpXrI2tdQpQEVwlQ" title="单细胞轨迹分析知多少--拟时间分析比较">单细胞轨迹分析知多少--拟时间分析比较</a>》中我们介绍了45种单细胞轨迹推断分析软件方法在以下4个方面的比较：</p>
<ol>
<li>准确性</li>
<li>可扩展性</li>
<li>稳定性</li>
<li>可用性</li>
</ol>
<p><strong><code>得出了几项重要结论</code></strong>：</p>
<blockquote>
<ul>
<li>轨迹推断（TI，trajectory inference）软件方法输入和输出接口的标准化对于TI方法的广泛应用非常关键</li>
<li>不同TI方法的组合存在一定的互补性，能够高概率的得出与实际生物学模型相符的轨迹推断结果</li>
<li>在1个项目的数据集上尝试多种TI分析算法，一是可以提供多种算法证据支持，二是如果多种算法的结果有冲突有利于我们深度思考可能的具体生物学原因</li>
</ul>
</blockquote>
<p>所以<a href="https://www.nature.com/articles/s41587-019-0071-9" title="A comparison of single-cell trajectory inference methods">A comparison of single-cell trajectory inference methods</a>(Saelens et al., 2019)的作者做了一个R包--<code>dyno</code>为终端用户提供完整的TI分析流程，<code>dyno特点</code>如下：</p>
<ol>
<li>统一59种TI方法的输入输出接口</li>
<li>提供交互式指南工具，可帮助用户选择最合适的TI方法</li>
<li>简化了轨迹的解释和可视化，包括根据基因表达或者cluster着色</li>
<li>还可以进行下游分析，例如潜在marker gene 的鉴定</li>
</ol>
<h1><code>dyno</code>安装</h1>
<p>dyno是一个R包，需要R/Rstudio运行环境，目前的存放地址是在<code>github</code>- <code>https://github.com/dynverse/dyno</code>，所以我们用devtools进行安装，打开Rstudio，输入以下命令：</p>
<div class="highlight"><pre><span></span>#install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;dynverse/dyno&quot;)
</pre></div>


<p>在这个过程中会安装一系列dynverse R包(dynwrap, dynplot, dynmethods, …)。</p>
<p>如果是在linux系统使用，需要安装<code>udunits</code>和<code>ImageMagick</code></p>
<ul>
<li>Debian / Ubuntu / Linux Mint: 
  <code>sudo apt-get install libudunits2-dev imagemagick</code></li>
<li>Fedora / CentOS / RHEL: 
  <code>sudo dnf install udunits2-devel ImageMagick-c++-devel</code></li>
</ul>
<p>运行TI方法需要用到<code>Docker</code>或者<code>Singularity</code> (version ≥ 3.0)，如果电脑系统是Windows或者MacOS，推荐使用docker，如果是在集群，推荐使用 Singularity。</p>
<p>如果是win10系统，可以运行<a href="https://hub.docker.com/editions/community/docker-ce-desktop-windows" title="Docker CE">Docker CE</a>。Singularity的安装请参考<code>https://www.sylabs.io/docs/</code>，如果已经安装了anaconda，可以执行<code>conda create -n singularity -c conda-forge singularity=3.0.1</code>安装</p>
<h2>安装过程遇到的问题</h2>
<h3>API rate limit exceeded</h3>
<p>因为dyno存放在了Github，在dyno的安装过程会使用<code>GitHub API</code>，默认的API限制是60次请求，所以会遇到<code>API rate limit exceeded</code>的问题，解决方式如下：</p>
<ul>
<li>在Rstudio命令行执行 <code>usethis::browse_github_pat()</code>建议一个GitHub token</li>
<li>执行<code>usethis::edit_r_environ()</code> 添加以下环境变量<code>GITHUB_PAT = 'your_github_token</code></li>
<li>重启R/Rstudio (重新载入 GITHUB_PAT) </li>
<li>再次重新安装:<code>devtools::install_github("dynverse/dyno")</code></li>
</ul>
<h3>其他安装问题</h3>
<p>如果没有<code>devtools</code>是不能安装<code>dyno</code>的，所以有些时候<code>devtools</code>的安装也会出现一些问题，例如网络不稳定等，这时可以采用国内的源进行安装：<code>install.packages('devtools', repos='https://mirrors.tuna.tsinghua.edu.cn/CRAN/')</code></p>
<p>在linux集群，如果没有<code>root</code>权限，安装使用<code>dyno</code>会非常困难，我准备了解决方案会在文章最后一节介绍。</p>
<h3>windows10/MACOS用户快速体验方式</h3>
<p>dyno的整个安装过程会非常漫长，如果是windows10或者MACOS用户可以在安装docker之后用我构建的R镜像：<code>seqyuan/seqyuan-r:v0.0.1</code>（R version 3.6.1）进行体验，这个R镜像安装了dyno、tidyverse及<code>Seurat</code>等包，可以免去安装过程快速体验dyno的强大，具体使用会在文章最后一节介绍。</p>
<h1>dyno的使用</h1>
<p><a href="https://github.com/dynverse/dyno" title="dyno">dyno</a>的官网<code>https://dynverse.org/</code>有详细的使用步骤介绍，在这里我们就不再重复官网测试数据的结果，重点介绍<code>Seurat</code>分析的10X单细胞转录组数据的结果怎么样和dyno对接。</p>
<h2>主要步骤</h2>
<p><a href="https://dynverse.org" title="dynverse">主要步骤如下</a></p>
<h4>准备数据</h4>
<h4>为数据集选择最佳方法</h4>
<h4>运行TI方法</h4>
<h4>轨迹可视化</h4>
<h4>从生物学角度解释轨迹</h4>
<ul>
<li>Rooting</li>
<li>Milestone labelling</li>
</ul>
<h4>基因表达在轨迹上的展示</h4>
<ul>
<li>A global overview of the most predictive genes</li>
<li>Lineage/branch markers</li>
<li>Genes important at bifurcation points</li>
</ul>
<p><img alt="46ad4bd28656eb7ac2b75bcc55c0d335.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_1.png"></p>
<h2>准备数据</h2>
<p>首先打开Rstudio（如果是linux，打开R命令行），载入以下R包</p>
<div class="highlight"><pre><span></span><span class="kn">library</span><span class="p">(</span>dyno<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>tidyverse<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>Matrix<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>Seurat<span class="p">)</span>
</pre></div>


<p>dyno对数据的包装主要通过<a href="https://dynverse.org/reference/dynwrap/">dynwrap</a>包来实现</p>
<h3>Gene expression data</h3>
<p>dynwrap要求raw counts和normalised(log2)表达数据，低表达的细胞、双细胞、坏细胞应该被提前过滤。使用Seurat处理后的数据一般都包含了这些步骤。</p>
<p><code>wrap_expression</code>要求raw counts和normalised表达为 sparse matrix (dgCMatrix）(<code>列</code>为genes/features，<code>行</code>为细胞) </p>
<div class="highlight"><pre><span></span><span class="c1">#读入seurat处理后的rds文件</span>
sdata <span class="o">&lt;-</span> <span class="kp">readRDS</span><span class="p">(</span>file <span class="o">=</span> <span class="s">&quot;/Users/yuanzan/Desktop/tmp/seurat_project.rds&quot;</span><span class="p">)</span>

<span class="c1">#添加raw counts和normalised expression</span>
<span class="c1">#seurat的矩阵需要进行行列转换，以使行为细胞，列为基因</span>
dataset <span class="o">&lt;-</span> wrap_expression<span class="p">(</span>
  counts <span class="o">=</span> <span class="kp">t</span><span class="p">(</span>sdata<span class="o">@</span>assays<span class="o">$</span>RNA<span class="o">@</span>counts<span class="p">),</span>
  expression <span class="o">=</span> <span class="kp">t</span><span class="p">(</span>sdata<span class="o">@</span>assays<span class="o">$</span>RNA<span class="o">@</span>data<span class="p">)</span>
<span class="p">)</span>

<span class="c1">#添加先验信息，这里添加的是开始转换的“细胞id”，后期可视化可以根据具体的轨迹推断结果进行调整</span>
dataset <span class="o">&lt;-</span> add_prior_information<span class="p">(</span>
  dataset<span class="p">,</span>
  start_id <span class="o">=</span> <span class="s">&quot;TTTGTTGCAACTCATG.wt&quot;</span>
<span class="p">)</span>

<span class="c1">#添加数据的cluster信息，这里我们直接用“seurat_clusters”即可</span>
dataset <span class="o">&lt;-</span> add_grouping<span class="p">(</span>
   dataset<span class="p">,</span>
   sdata<span class="o">$</span>seurat_clusters
<span class="p">)</span>
</pre></div>


<h2>为数据集选择最佳方法</h2>
<p>如果是MACOS，可以执行以下命令以获取推荐的最佳TI方法</p>
<div class="highlight"><pre><span></span>guidelines <span class="o">&lt;-</span> guidelines_shiny<span class="p">(</span>dataset<span class="p">)</span>
methods_selected <span class="o">&lt;-</span> guidelines<span class="o">$</span>methods_selected
</pre></div>


<p>执行<code>guidelines &lt;- guidelines_shiny(dataset)</code>命令之后会弹出一个<code>shiny</code>页面，如下图
<img alt="519ae00492e91fe380b9cdd350ab163f.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_2.png">
因为我们给了<code>dataset</code>，所以guidelines页面会自动填写细胞数和基因数参数，我们再把预期的运行时间和运行内存选择设置一下，可以扩大选择TI方法的选择范围。</p>
<p><img alt="c2b214fdcd4d387d86d2ac34ea4ebf25.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_3.png"></p>
<p>通过选择对于推断轨迹的预期以及左侧关于内存、运行时间等各项参数，<code>guidelines</code>为我推荐了几个TI算法的组合，点击右上角的<code>Close &amp; use</code>关闭这个shiny页面。
<img alt="6d7838df846723fc2d91e16daad883a8.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_4.png"></p>
<p><code>methods_selected &lt;- guidelines$methods_selected</code>这个命令是把guidelines推荐给我们的方法赋值给<code>methods_selected</code>，这是一个字符串向量。
<img alt="9644a95b042cec2e1edd20de5d836535.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_5.png">
如果我们是在linux集群上进行的操作，这一步弹出guidelines shiny的界面可能需要通过很复杂的设置才能获得，而且也不利于把命令行写成脚本执行。幸好我们有替代方法，那就是</p>
<ol>
<li>通过命令<code>dim(dataset$counts)</code>获取cell number和feature/gene number</li>
<li>打开网址<code>http://guidelines.dynverse.org</code></li>
<li>在guidelines网页填写细胞数、gene数以及其他设置</li>
<li>打对勾的TI方法就是推荐给我们的，如下图
<img alt="ca6a6e1e4eb26018cf28f1f5d64430a3.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_6.png"></li>
</ol>
<h2>运行TI方法</h2>
<div class="highlight"><pre><span></span>model_slingshot <span class="o">&lt;-</span> infer_trajectory<span class="p">(</span>dataset<span class="p">,</span> methods_selected<span class="p">[</span><span class="m">1</span><span class="p">])</span>
model_paga_tree <span class="o">&lt;-</span> infer_trajectory<span class="p">(</span>dataset<span class="p">,</span> methods_selected<span class="p">[</span><span class="m">2</span><span class="p">])</span>
model_paga <span class="o">&lt;-</span> infer_trajectory<span class="p">(</span>dataset<span class="p">,</span> methods_selected<span class="p">[</span><span class="m">3</span><span class="p">])</span>
model_mst <span class="o">&lt;-</span> infer_trajectory<span class="p">(</span>dataset<span class="p">,</span> methods_selected<span class="p">[</span><span class="m">4</span><span class="p">])</span>
</pre></div>


<blockquote>
<p>作者把相应TI方法的源码封装在了docker镜像里，调用相应方法的后台操作其实是：先拉取TI方法对应的镜像，然后再R进程内部启动docker容器执行算法，所以如果你使用的方法事先没有经过<code>docker pull</code>，那么算法的执行时间其实包含了拉取镜像所需要的时间，建议把常用到的TI方法先pull到本地，这样就能使脚本代码的执行更快。</p>
</blockquote>
<p>目前作者封装的59种TI方法列表见网址：https://github.com/dynverse/dynmethods#list-of-included-methods，用到代码中的调用一般是单词的所有字母都小写，单词之间加下划线。</p>
<p>guidelines推荐给我的<code>methods_selected</code>包含4个TI方法，本文开头介绍了《<a href="https://mp.weixin.qq.com/s/6ZVTR3wpXrI2tdQpQEVwlQ" title="单细胞轨迹分析知多少--拟时间分析比较">单细胞轨迹分析知多少--拟时间分析比较</a>》得出了几项重要结论，理论指导实践，所以我打算把4中TI方法都试一下，看看结果怎么样。</p>
<h3>简单可视化</h3>
<div class="highlight"><pre><span></span>model <span class="o">&lt;-</span> model_slingshot
plot_dimred<span class="p">(</span>
  model<span class="p">,</span> 
  expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span> 
  grouping <span class="o">=</span> dataset<span class="o">$</span>grouping
<span class="p">)</span>
</pre></div>


<p>slingshot
<img alt="e63dc6bb7e6eeae9dd5dcad74a2794b5.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_7.png">
paga
<img alt="31a6b4ef3567abdc51300a2a714f107b.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_8.png">
mst
<img alt="730af4534faea2ef6800f6cdc7f08328.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_9.png">
paga_tree
<img alt="a4212e00ffd8a6bdaa783bc6099da24f.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_10.png"></p>
<h2>从生物学角度解释轨迹</h2>
<h3>Rooting</h3>
<p>大多数方法没有直接的方法来推断轨迹的方向性。在这种情况下，应该使用一些外部信息来“确定”轨迹，例如使用一组marker基因</p>
<p>添加rooting gene列表，这里需要根据具体的生物学问题来定，为了更快速的演示，我不再深究，这里仅给一个示例作为演示用</p>
<div class="highlight"><pre><span></span><span class="c1"># 初步来看我认为slingshot方法得到的结果和我的预期比较相近，后续将以此方法得到的model进行后续演示</span>
model <span class="o">&lt;-</span> model_slingshot
model <span class="o">&lt;-</span> model <span class="o">%&gt;%</span> 
  add_root_using_expression<span class="p">(</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;KLRB1&quot;</span><span class="p">),</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">)</span>
</pre></div>


<h3>Milestone labelling</h3>
<p>可以使用marker基因标记里程碑，这些标签随后可用于后续分析和可视化</p>
<div class="highlight"><pre><span></span><span class="c1"># 通过model$milestone_ids我们知道了有4个里程碑</span>
<span class="c1"># 这个示例我们标记3个里程碑，此处仅为示例，请忽略标签意义</span>
model <span class="o">&lt;-</span> label_milestones_markers<span class="p">(</span>
  model<span class="p">,</span>
  markers <span class="o">=</span> <span class="kt">list</span><span class="p">(</span>
    Tn <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;CCR7&quot;</span><span class="p">),</span>
    NK <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;KLRD1&quot;</span><span class="p">),</span>
    CTL <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;GZMK&quot;</span><span class="p">)</span>
  <span class="p">),</span>
  dataset<span class="o">$</span><span class="kp">expression</span>
<span class="p">)</span>
</pre></div>


<h2>轨迹可视化</h2>
<p>dyno提供了降维、轨迹模型和细胞聚类等可视化展示方法</p>
<h3>细胞亚群在轨迹上的展示</h3>
<div class="highlight"><pre><span></span>model <span class="o">&lt;-</span> model <span class="o">%&gt;%</span> add_dimred<span class="p">(</span>dyndimred<span class="o">::</span>dimred_mds<span class="p">,</span> expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">)</span>
plot_dimred<span class="p">(</span>
  model<span class="p">,</span> 
  expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span> 
  grouping <span class="o">=</span> dataset<span class="o">$</span>grouping<span class="p">,</span>
  label_milestones <span class="o">=</span> <span class="kc">TRUE</span>
<span class="p">)</span>
</pre></div>


<p><img alt="30373eae7fa26a5d3bb047e28d07cb70.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_11.png"></p>
<p>通过基础的4个细胞亚群的轨迹分布，我们明显的看到3亚群分为了2部分，所以我打算后续调整第3群为2个亚群再做轨迹分析，以便更好的体现它们的差异。</p>
<p>我们设置了<code>label_milestones = TRUE</code>，所以里程碑被标注到了轨迹模型上，model一共有4个里程碑我们只标注了3个所以上图只显示3个。
<img alt="17e7638ced15f8eab0b74d937770e53a.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_12.png"></p>
<h3>基因表达在轨迹上的展示</h3>
<p>这里我们做CCR7和CCL5两个基因表达示例</p>
<div class="highlight"><pre><span></span>plot_dimred(
  model, 
  expression_source = dataset$expression, 
  color_cells = &quot;feature&quot;,
  feature_oi = &quot;CCR7&quot;,
  color_density = &quot;grouping&quot;,
  grouping = dataset$grouping,
  label_milestones = TRUE
)
</pre></div>


<p><img alt="d755b128fea3205a3a9e43511e316887.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_13.png">
<img alt="8a01d2caeb6be4c9590196d2f3924e5e.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_14.png"></p>
<p>在展示基因表达的同时，不同细胞亚群以细胞的背景颜色表示，以下官方示例展示的更清楚一些
<img alt="cb0a43f5628a0e47bd86b0a7c1db5d2b.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_15.png"></p>
<h2>预测和可视化感兴趣的基因</h2>
<p>dyno整合了几种方法来从轨迹中提取候选标记基因/特征</p>
<h3>最具预测性的基因的全球概述</h3>
<p>默认情况下，绘制热图时会计算总体上最重要的基因</p>
<div class="highlight"><pre><span></span>plot_heatmap<span class="p">(</span>
  model<span class="p">,</span>
  expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span>
  grouping <span class="o">=</span> dataset<span class="o">$</span>grouping<span class="p">,</span>
  features_oi <span class="o">=</span> <span class="m">50</span>
<span class="p">)</span>
</pre></div>


<p><img alt="1481b75d008fed605270e71e36cec8bc.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_16.png"></p>
<h3>Lineage/branch markers</h3>
<p>dyno还有提取特定分支的功能，例如：当细胞分化成神经元时改变的基因</p>
<div class="highlight"><pre><span></span>branch_feature_importance <span class="o">&lt;-</span> calculate_branch_feature_importance<span class="p">(</span>model<span class="p">,</span> expression_source<span class="o">=</span>dataset<span class="o">$</span><span class="kp">expression</span><span class="p">)</span>

select_features <span class="o">&lt;-</span> branch_feature_importance <span class="o">%&gt;%</span> 
  filter<span class="p">(</span>to <span class="o">==</span> <span class="kp">which</span><span class="p">(</span>model<span class="o">$</span>milestone_labelling <span class="o">==</span><span class="s">&quot;CTL&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span> 
  top_n<span class="p">(</span><span class="m">30</span><span class="p">,</span> importance<span class="p">)</span> <span class="o">%&gt;%</span> 
  pull<span class="p">(</span>feature_id<span class="p">)</span>

plot_heatmap<span class="p">(</span>
  model<span class="p">,</span> 
  expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span> 
  features_oi <span class="o">=</span> select_features
<span class="p">)</span>
</pre></div>


<p><img alt="793e06efdff94d4fa90b0bd3132b39c5.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_17.png"></p>
<h3>分叉点重要的基因</h3>
<p>提取在分支点发生变化的特征/基因</p>
<div class="highlight"><pre><span></span>branching_milestone <span class="o">&lt;-</span> model<span class="o">$</span>milestone_network <span class="o">%&gt;%</span> group_by<span class="p">(</span>from<span class="p">)</span> <span class="o">%&gt;%</span> filter<span class="p">(</span>n<span class="p">()</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">)</span> <span class="o">%&gt;%</span> pull<span class="p">(</span>from<span class="p">)</span> <span class="o">%&gt;%</span> first<span class="p">()</span>

branch_feature_importance <span class="o">&lt;-</span> calculate_branching_point_feature_importance<span class="p">(</span>model<span class="p">,</span> expression_source<span class="o">=</span>dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span> milestones_oi <span class="o">=</span> branching_milestone<span class="p">)</span>

branching_point_features <span class="o">&lt;-</span> branch_feature_importance <span class="o">%&gt;%</span> top_n<span class="p">(</span><span class="m">20</span><span class="p">,</span> importance<span class="p">)</span> <span class="o">%&gt;%</span> pull<span class="p">(</span>feature_id<span class="p">)</span>

plot_heatmap<span class="p">(</span>
  model<span class="p">,</span>
  expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span>
  features_oi <span class="o">=</span> branching_point_features
<span class="p">)</span>
</pre></div>


<p><img alt="7ee4adbc0789dca82102908e6c448b6f.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_18.png"></p>
<div class="highlight"><pre><span></span>space <span class="o">&lt;-</span> dyndimred<span class="o">::</span>dimred_mds<span class="p">(</span>dataset<span class="o">$</span><span class="kp">expression</span><span class="p">)</span>
map<span class="p">(</span>branching_point_features<span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">],</span> <span class="kr">function</span><span class="p">(</span>feature_oi<span class="p">)</span> <span class="p">{</span>
  plot_dimred<span class="p">(</span>model<span class="p">,</span> dimred <span class="o">=</span> space<span class="p">,</span> expression_source <span class="o">=</span> dataset<span class="o">$</span><span class="kp">expression</span><span class="p">,</span> feature_oi <span class="o">=</span> feature_oi<span class="p">,</span> label_milestones <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span> <span class="o">+</span>
    theme<span class="p">(</span>legend.position <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> <span class="o">+</span>
    ggtitle<span class="p">(</span>feature_oi<span class="p">)</span>
<span class="p">})</span> <span class="o">%&gt;%</span> patchwork<span class="o">::</span>wrap_plots<span class="p">()</span>
</pre></div>


<p><img alt="d383855a6a239ea4944257d23db8f26d.png" src="https://raw.githubusercontent.com/seqyuan/blog/master/images/dyno/dyno_19.png"></p>
<h1>快速免安装体验方式</h1>
<p>对于MACOS或者有root权限的用户，如果想快速体验dyno，可以在安装并启动docker之后，执行以下命令行进入docker环境命令行，然后打开<code>R</code>终端。</p>
<div class="highlight"><pre><span></span>sudo docker run -it --memory=20G --memory-swap=20G --oom-kill-disable -v /home/zanyuan:/zanyuan -v /tmp:/tmp -v /usr/bin/:/dockerbin -v /var/run/docker.sock:/var/run/docker.sock seqyuan/seqyuan-r:v0.0.1 /bin/bash
</pre></div>


<p>理解上面的代码可能需要一些docker方面的知识，其中<code>-v /home/zanyuan:/zanyuan</code> 这一行指的是本地的存储挂载为docker容器里的<code>/zanyuan</code>地址；<code>-v /tmp:/tmp</code>是临时存储目录挂载；<code>-v /usr/bin/:/dockerbin</code>指的是docker所在本地路径挂载为容器里的<code>/dockerbin</code>；<code>seqyuan/seqyuan-r:v0.0.1</code>这个是docker启动的镜像名称，我构建的<a href="https://hub.docker.com/repository/docker/seqyuan/seqyuan-r" title="dockerfile">dockerfile</a>请见docker hub官网。</p>
<p>如果运行时间比较长，可以把代码写到一个名为<code>test_dyno.R</code>脚本里例如：</p>
<div class="highlight"><pre><span></span><span class="kn">library</span><span class="p">(</span>dyno<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>tidyverse<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>Matrix<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>Seurat<span class="p">)</span>

pso <span class="o">&lt;-</span> <span class="kp">readRDS</span><span class="p">(</span>file <span class="o">=</span> <span class="s">&quot;/zanyuan/data/cd8_t1.rds&quot;</span><span class="p">)</span>

dataset <span class="o">&lt;-</span> wrap_expression<span class="p">(</span>
  counts <span class="o">=</span> <span class="kp">t</span><span class="p">(</span>pso<span class="o">@</span>assays<span class="o">$</span>RNA<span class="o">@</span>counts<span class="p">),</span>
  expression <span class="o">=</span> <span class="kp">t</span><span class="p">(</span>pso<span class="o">@</span>assays<span class="o">$</span>RNA<span class="o">@</span>data<span class="p">)</span>
<span class="p">)</span>


dataset <span class="o">&lt;-</span> add_prior_information<span class="p">(</span>
  dataset<span class="p">,</span>
  start_id <span class="o">=</span> <span class="s">&quot;AGCGTCGTCCCTAACC.5&quot;</span>
<span class="p">)</span>

model <span class="o">&lt;-</span> infer_trajectory<span class="p">(</span>dataset<span class="p">,</span> <span class="s">&quot;paga&quot;</span><span class="p">)</span>

<span class="kp">saveRDS</span><span class="p">(</span>model<span class="p">,</span> file <span class="o">=</span> <span class="s">&quot;/zanyuan/data/cd8_paga.model&quot;</span><span class="p">,</span> ascii <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> version <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span> compress <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> refhook <span class="o">=</span> <span class="kc">NULL</span><span class="p">)</span>
</pre></div>


<p>然后执行</p>
<div class="highlight"><pre><span></span>sudo docker run -it --memory=20G --memory-swap=20G --oom-kill-disable -v /home/zanyuan:/zanyuan -v /tmp:/tmp -v /usr/bin/:/dockerbin -v /var/run/docker.sock:/var/run/docker.sock seqyuan/seqyuan-r:v0.0.1 Rscript /zanyuan/test_ti.R
</pre></div>


<h1>后续</h1>
<p>后续会推出在linux集群无root权限的情况下调用singularity来执行dyno</p>
            
            
            <hr/>
        </div>
        <section>
        <div class="span2" style="float:right;font-size:0.9em;">
            <h4><a href="/author/seqyuan.html"> <i class="icon-user"></i>seqyuan</a></h4>

            <!--<h4>Published</h4>-->
			<h4><time pubdate="pubdate" datetime="2020-01-13T00:00:00+08:00"><i class="icon-calendar"></i>2020-01-13</time></h4>

            <h4><i class="icon-folder-open"></i>Category</h4>
            <a class="category-link" href="/categories.html#single-cell-ref">single-cell</a>
            <h4><i class="icon-tag"></i>Tags</h4>
            <ul class="list-of-tags tags-in-article">
                <li><a href="/tags.html#pseudotime-trajectory-inference-ref">pseudotime， trajectory inference
                    <span>2</span>
</a></li>
            </ul>
        </div>
        </section>
</div>
</article>
                </div>
                <div class="span1"></div>
            </div>
        </div>
        <div id="push"></div>
    </div>
<footer>
<div id="footer">
    <ul class="footer-content">
        <li class="elegant-power">Designed by <a href="http://www.seqyuan.com/" title="Seqyuan Home Page">Yuan Zan</a></li>
    </ul>
</div>
</footer>            <script src="http://code.jquery.com/jquery.min.js"></script>
        <script src="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/js/bootstrap.min.js"></script>
        <script>
            function validateForm(query)
            {
                return (query.length > 0);
            }
        </script>

    
    </body>
    <!-- Theme: Elegant built for Pelican
    License : http://oncrashreboot.com/pelican-elegant -->
</html>